Problem Japan symbols and attachment name

### system info

Haraka | Haraka.js — Version: 2.8.24
 --- | :--- 
Node | v10.16.0
OS | Linux hostname 4.4.0-146-generic #172-Ubuntu SMP Wed Apr 3 09:00:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
openssl | OpenSSL 1.1.0h  27 Mar 2018 (Library: OpenSSL 1.1.1b  26 Feb 2019)

### Expected behavior

Correct enconditng of message and correct name of attachment expected

### Observed behavior

We received complaints that the message does not contain the correct characters and attachment name. 
Upon further examination, it became clear that haraka could not decipher the body of the message and jumbled the characters.
The attachment's name is collected incorrectly and cannot produce decodeURIComponent as
filename*0*, filename*1* are the main components of the attachment's name. 

### Steps to reproduce

```javascript
exports.hook_rcpt_ok = (next, connection) => {
    connection.transaction.parse_body = true;
    
    connection.transaction.attachment_hooks(function (contentType, filename, body, stream) {
        connection.loginfo(filename);
    });
    
    next(OK);
};


exports.hook_data_post = function (next, connection, params) {
    connection.loginfo(
        connection.transaction.body.children[0]
    );

    next();
};
```

and send this email

```
To: example@example.com
From: example@example.com
Subject: Example
Date: Wed, 17 Jul 2019 07:02:19 +0900
User-Agent: Mozilla/5.0 (Windows NT 6.3; WOW64; rv:60.0) Gecko/20100101
 Thunderbird/60.8.0
MIME-Version: 1.0
Content-Type: multipart/mixed;
 boundary="------------CDA77F30E7B3942A1340A111"
Content-Language: en-US



This is a multi-part message in MIME format.
--------------CDA77F30E7B3942A1340A111
Content-Type: multipart/alternative;
 boundary="------------66C1E7909E2B8C6834E8B83E"


--------------66C1E7909E2B8C6834E8B83E
Content-Type: text/plain; charset=utf-8; format=flowed
Content-Transfer-Encoding: 8bit

どうぞ宜しくお願い申し上げます。



--------------66C1E7909E2B8C6834E8B83E
Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: 8bit

<html>
  <head>

    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  </head>
  <body text="#000000" bgcolor="#FFFFFF">
    </tt><tt> どうぞ宜しくお願い申し上げます。</tt><tt><br>
  </body>
</html>

--------------66C1E7909E2B8C6834E8B83E--

--------------CDA77F30E7B3942A1340A111
Content-Type: plain/text;
Content-Transfer-Encoding: base64
Content-Disposition: attachment;
 filename*0*=utf-8''%E8%AC%9B%E6%BC%94%E4%BC%9A%E6;
 filename*1*=%A1%88%E5%86%85%E6%9B%B8%EF%BC%86%E7%94%B3%E8%BE%BC%E6%9B%B8;
 filename*2*=%E6%94%B9%2Etxt

dGVzdA==
--------------CDA77F30E7B3942A1340A111--
```

As a result we see the following problem

by attachment
```
2019-07-19T14:46:16.907Z [ERROR] [-] [core] Decode header failed: filename*0*: %E8%AC%9B%E6%BC%94%E4%BC%9A%E6
2019-07-19T14:46:16.907Z [ERROR] [-] [core] Decode header failed: filename*1*: %A1%88%E5%86%85%E6%9B%B8%EF%BC%86%E7%94%B3%E8%BE%BC%E6%9B%B8
2019-07-19T14:46:16.908Z [INFO] [4FF427EE-1DCB-4942-A10D-9556BD05E93A.1] [core] %E8%AC%9B%E6%BC%94%E4%BC%9A%E6%A1%88%E5%86%85%E6%9B%B8%EF%BC%86%E7%94%B3%E8%BE%BC%E6%9B%B8改.txt
```

by mail body
```
       bodytext: 'iF^WOJXD3W\nR~Y\u0002\n\n\n\n',
       body_text_encoded: 'どうぞ宜しくお願い申し上げます。\n\n\n\n',

```

This issue regarding the message body is similar to my previous issue https://github.com/haraka/Haraka/issues/2548
The issue with the attachment is a new one.

--- a/Changes.md
+++ b/Changes.md
@@ -19,6 +19,7 @@
 
 * mailheader: fully quality header name in _remove_more #2647
 * haraka: Connection.createConnection is not a constructor #2816
+* problems with japanese characters in body and part header #2675
 
 
 ## 2.8.24 - Mar 12, 2019

--- a/mailbody.js
+++ b/mailbody.js
@@ -227,12 +227,7 @@ class Body extends events.EventEmitter {
         }
 
         if (/UTF-?8/i.test(enc)) {
-            if (this.decode_function === this.decode_8bit) {
-                // source string was UTF-8 but parsed as binary
-                this.bodytext = buf.toString('binary');
-            } else {
-                this.bodytext = buf.toString();
-            }
+            this.bodytext = buf.toString();
             return;
         }
 
@@ -368,6 +363,9 @@ class Body extends events.EventEmitter {
     }
 
     decode_8bit (line) {
+        if (/UTF-?8/i.test(this.body_encoding)) {
+            return Buffer.from(line, 'utf-8');
+        }
         return Buffer.from(line, 'binary');
     }
 }

--- a/mailheader.js
+++ b/mailheader.js
@@ -64,21 +64,9 @@ class Header {
             cur_lang: '', // Secondary languages are ignored for our purposes
         };
 
-        _decode_rfc2231(rfc2231_params, val);
+        val = _decode_rfc2231(rfc2231_params, val);
 
         // console.log(rfc2231_params);
-
-        for (const key in rfc2231_params.keys) {
-            val = val + ' ' + key + '="';
-            /* eslint no-constant-condition: 0 */
-            for (let i=0; true; i++) {
-                const _key = key + '*' + i;
-                const _val = rfc2231_params.kv[_key];
-                if (_val === undefined) break;
-                val = val + _val;
-            }
-            val = val + '";';
-        }
 
         // strip 822 comments in the most basic way - does not support nested comments
         // val = val.replace(/\([^\)]*\)/, '');
@@ -229,6 +217,34 @@ function _decode_header (matched, encoding, lang, cte, data) {
 }
 
 function _decode_rfc2231 (params, str) {
+    _parse_rfc2231(params, str);
+
+    for (const key in params.keys) {
+        str = str + ' ' + key + '="';
+        /* eslint no-constant-condition: 0 */
+        let merged = '';
+        for (let i=0; true; i++) {
+            const _key = key + '*' + i;
+            const _val = params.kv[_key];
+            if (_val === undefined) break;
+            merged = merged + _val;
+        }
+
+        try {
+            merged = decodeURIComponent(merged);
+        }
+        catch (e) {
+            logger.logerror("Decode header failed: " + key + ": " + merged);
+        }
+        merged = params.cur_enc ? try_convert(merged, params.cur_enc) : merged;
+
+        str = str + merged + '";';
+    }
+
+    return str;
+}
+
+function _parse_rfc2231 (params, str) {
     /*
     To explain the regexp below, the params are:
 
@@ -272,7 +288,7 @@ function _decode_rfc2231 (params, str) {
     const lang_match = /^(.*?)'(.*?)'(.*)/.exec(value);
     if (lang_match) {
         if (key_actual == params.cur_key && lang_match[2] != params.cur_lang) {
-            return _decode_rfc2231(params, str); // same key, different lang, throw it away
+            return _parse_rfc2231(params, str); // same key, different lang, throw it away
         }
         params.cur_enc = lang_match[1];
         params.cur_lang = lang_match[2];
@@ -285,12 +301,6 @@ function _decode_rfc2231 (params, str) {
 
     params.cur_key = key_actual;
     params.keys[key_actual] = '';
-    try {
-        value = decodeURIComponent(value);
-    }
-    catch (e) {
-        logger.logerror("Decode header failed: " + key + ": " + value);
-    }
-    params.kv[key_actual + '*' + key_id] = params.cur_enc ? try_convert(value, params.cur_enc) : value;
-    return _decode_rfc2231(params, str); // Get next one
+    params.kv[key_actual + '*' + key_id] = value;
+    return _parse_rfc2231(params, str); // Get next one
 }

--- a/tests/mailbody.js
+++ b/tests/mailbody.js
@@ -85,6 +85,7 @@ exports.basic = {
             ['iso-8859-2', '8-bit', "\x50\xF8\x69\x68\x6C\x61\xB9\x6F\x76\x61\x63\xED\x20\xFA\x64\x61\x6A\x65\x0A", "Přihlašovací údaje\n"],
             ['iso-8859-2', 'quoted-printable', "P=F8ihla=B9ovac=ED =FAdaje\n", "Přihlašovací údaje\n"],
             ['iso-8859-2', 'base64', "UPhpaGxhuW92YWPtIPpkYWplCgo=\n", "Přihlašovací údaje\n\n"],
+            ['utf-8', '8-bit', "どうぞ宜しくお願い申し上げます", "どうぞ宜しくお願い申し上げます"]
         ];
 
         test.expect(tests.length);

--- a/tests/mailheader.js
+++ b/tests/mailheader.js
@@ -98,3 +98,18 @@ exports.remove = {
         test.done()
     }
 }
+
+exports.decode = {
+    'multiline 8bit header (#2675)': test => {
+        this.h = new Header();
+        this.h.parse ([
+            "Content-Disposition: attachment;\n",
+            " filename*0*=utf-8''%E8%AC%9B%E6%BC%94%E4%BC%9A%E6;\n",
+            " filename*1*=%A1%88%E5%86%85%E6%9B%B8%EF%BC%86%E7%94%B3%E8%BE%BC%E6%9B%B8;\n",
+            " filename*2*=%E6%94%B9%2Etxt\n"
+        ]);
+        console.log(this.h.get_decoded('content-disposition'));
+        test.ok(this.h.get_decoded('content-disposition').includes('講演会案内書＆申込書改.txt'));
+        test.done();
+    }
+}

